readinglist/resource.py

```diff
@@ -270,11 +282,110 @@ def _extract_sorting(self, queryparams):
                     }
                     self.raise_invalid(**error_details)
 
+                if field == 'last_modified':
+                    last_modified_used = True
+
                 direction = -1 if order == '-' else 1
                 sorting.append((field, direction))
 
+        if not last_modified_used:
+            sorting.append(('last_modified', -1))
         return sorting
 
+    def _extract_pagination_token(self, queryparams):
+        """Get pagination params."""
+        settings = self.request.registry.settings
+        try:
+            limit = int(queryparams.get('_limit', settings.get(
+                'readinglist.paginate_by', DEFAULT_PAGINATE_BY)))
+        except ValueError:
+            error_details = {
+                'name': None,
+                'location': 'querystring',
+                'description': "_limit should be an integer"
+            }
+            self.request.errors.add(**error_details)
+            raise errors.json_error(self.request.errors)
+
+        token = queryparams.get('_token', None)
+        filters = []
+        if token:
+            rules = json.loads(b64decode(token))
+            for rule in rules:
+                filters.append(self._extract_filters(rule))
+        return filters, limit
+
+    def _next_page_url(self, sorting, limit, last_record):
+        queryparams = self.request.GET.copy()
+        queryparams['_limit'] = limit
+        queryparams['_token'] = self._build_pagination_token(
+            sorting, last_record)
+        return '%s%s?%s' % (self.request.host_url, self.request.path_info,
+                            urlencode(queryparams))
+
+    def _build_pagination_token(self, sorting, last_record):
+        """Build a continuation filtering rules.
+
+        Something like that when sorting::
+
+            (status == 2 and last_modified < 123) OR status < 2
+
+
+        Something like that when sorting only per last_modified::
+
+            last_modified < 123
```

Comment:
Maybe it would help if you write a small summary of what the code does. I try here, to check that I understood well :)

```
Build a continuation token, as a piece of JSON encoded in base64, that will be
used by the client to resume the current pagination state.

The goal of the algorithm is the build the set of filters (using min/max/gt/lt) that
will reduce the dataset of the next page. The combination of sorting and paginating
is complex to resolve for this problem. Depending on the direction of sorting for each
field, the set of filters will target values above or below last record field values.

#. Identify if we sort on last modified (use ``_since`` or ``_to`` relatively)
#. Loop on the sorting fields and their direction
#. Build a list of filters for each of them using ``gt`` or ``lt`` depending on the direction
#. Serialize the filters and encode to base64

````